zbus: optional unique channel numeric identifiers

JordanYates opened the pull request:
Add the option for a unique numeric identifiers to be attached to a
zbus channel. This identifier can then be used to lookup the channel
structure at runtime. This is useful in three situations (that I can
immediately think of).

Firstly for external interaction, i.e a text shell or remote procedure
calls. The current state of a channel can be queried by an ID that never
changes, as opposed to the external entity needing to know the exact
memory address of the channel for a given application binary.

Secondly for integrating with dynamically loaded extensions (llext).
These extensions can hook into the existing data streams without each
individual channel needing to be exported and visible to the loader.

Thirdly, for common pieces of code that should compile regardless
of which channels are present in the application. Instead of getting
a link error when the channel fails to resolve, the user gets a NULL pointer
at runtime instead that can be handled/skipped.

The new macro duplicates the implementation of `ZBUS_CHAN_DEFINE` as
several of the macros parameters are lists that expand too many times
if passed through to a common implementation backend.

Documentation will be added if the concept is approved.

JordanYates committed:
zbus: optional unique channel numeric identifiers

Add the option for a unique numeric identifiers to be attached to a
zbus channel. This identifier can then be used to lookup the channel
structure at runtime. This is useful in two situations (that I can
immediately think of).

Firstly for external interaction, i.e a text shell or remote procedure
calls. The current state of a channel can be queried by an ID that never
changes, as opposed to the external entity needing to know the exact
memory address of the channel for a given application binary.

Secondly for integrating with dynamically loaded extensions (llext).
These extensions can hook into the existing data streams without each
individual channel needing to be exported and visible to the loader.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
tests: zbus: channel_id: test channel IDs

Test the new channel identifiers.

Signed-off-by: Jordan Yates <jordan@embeint.com>

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -74,6 +74,10 @@ struct zbus_channel {
 	/** Channel name. */
 	const char *const name;
+#endif
+#if defined(CONFIG_ZBUS_CHANNEL_ID) || defined(__DOXYGEN__)
+	/** Unique numeric channel identifier. */
+	uint32_t id;
 #endif
```
Please, add a line before and after to keep consistency. 
```suggestion

#if defined(CONFIG_ZBUS_CHANNEL_ID) || defined(__DOXYGEN__)
	/** Unique numeric channel identifier. */
	uint32_t id;
#endif

```

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -621,6 +676,20 @@ static inline const char *zbus_chan_name(const struct zbus_channel *chan)
+ */
+const struct zbus_channel *zbus_chan_from_id(uint32_t channel_id);
+
+#endif
```
```suggestion
#endif /* CONFIG_ZBUS_CHANNEL_ID */
```

rodrigopex reviewed (changes requested):
Thanks for the PR. That is a compelling feature. There are just some suggestions.

JordanYates commented on include/zephyr/zbus/zbus.h:
```diff
@@ -348,19 +358,64 @@ struct zbus_channel_observation {
+ * This macro defines a channel.
+ *
+ * @param _name The channel's name.
+ * @param _id The channel's unique numeric identifier.
+ * @param _type The Message type. It must be a struct or union.
+ * @param _validator The validator function.
+ * @param _user_data A pointer to the user data.
+ *
+ * @see struct zbus_channel
+ * @param _observers The observers list. The sequence indicates the priority of the observer. The
+ * first the highest priority.
+ * @param _init_val The message initialization.
+ */
+#define ZBUS_CHAN_ID_DEFINE(_name, _id, _type, _validator, _user_data, _observers, _init_val) \
+	static _type _CONCAT(_zbus_message_, _name) = _init_val;                          \
+	static struct zbus_channel_data _CONCAT(_zbus_chan_data_, _name) = {              \
+		.observers_start_idx = -1,                                                \
+		.observers_end_idx = -1,                                                  \
+		IF_ENABLED(CONFIG_ZBUS_PRIORITY_BOOST, (                                  \
+			.highest_observer_priority = ZBUS_MIN_THREAD_PRIORITY,            \
+		))                                                                        \
+	};                                                                                \
+	static K_MUTEX_DEFINE(_CONCAT(_zbus_mutex_, _name));                              \
+	_ZBUS_CPP_EXTERN const STRUCT_SECTION_ITERABLE(zbus_channel, _name) = {           \
+		ZBUS_CHANNEL_NAME_INIT(_name) /* Maybe removed */                         \
+		IF_ENABLED(CONFIG_ZBUS_CHANNEL_ID, (.id = _id,))                          \
+		.message = &_CONCAT(_zbus_message_, _name),                               \
+		.message_size = sizeof(_type),                                            \
+		.user_data = _user_data,                                                  \
+		.validator = _validator,                                                  \
+		.data = &_CONCAT(_zbus_chan_data_, _name),                                \
+		IF_ENABLED(ZBUS_MSG_SUBSCRIBER_NET_BUF_POOL_ISOLATION, (                  \
+			.msg_subscriber_pool = &_zbus_msg_subscribers_pool,               \
+		))                                                                        \
+	};                                                                                \
+	/* Extern declaration of observers */                                             \
+	ZBUS_OBS_DECLARE(_observers);                                                     \
+	/* Create all channel observations from observers list */                         \
+	FOR_EACH_FIXED_ARG_NONEMPTY_TERM(_ZBUS_CHAN_OBSERVATION, (;), _name, _observers)
+
```
As mentioned in the PR description, I haven't found a way to do this properly as the `_observers` and `_init_val` parameters are actually lists, and adding the extra indirection expands the list into extra parameters.
I could resolve this by changing `, _observers, _init_val` to `...` and  expanding it in the common macro. This would make the macro args less clear though.

JordanYates commented on subsys/zbus/zbus.c:
```diff
@@ -82,11 +82,52 @@ int _zbus_init(void)
+
+#if defined(CONFIG_ZBUS_CHANNEL_ID)
+		/* Check for duplicate channel IDs */
+		if (chan->id == ZBUS_CHAN_ID_INVALID) {
+			continue;
+		}
```
I disagree, if common libraries are creating internal zbus channels without ids the whole system should not be constrained to the lowest common denominator.

JordanYates commented on tests/subsys/zbus/channel_id/src/main.c:
```diff
@@ -0,0 +1,48 @@
+	zassert_is_null(zbus_chan_from_id(ZBUS_CHAN_ID_INVALID));
+
+	/* Standard retrieval */
+	zassert_equal(&chan_a, zbus_chan_from_id(CHAN_A));
+	zassert_equal(&chan_c, zbus_chan_from_id(CHAN_C));
+	zassert_equal(&chan_e, zbus_chan_from_id(CHAN_E));
+	zassert_equal(&chan_f, zbus_chan_from_id(CHAN_F));
```
It is not desired flexibility, but a consequence of the fact we have no way to ensure at compile-time there are no duplicate IDs.

It is therefore possible for the IDs to be duplicated, so some sort of behavior needs to be specified.
I have chosen "either channel with the ID will be returned".

rodrigopex replied:
I agree, but I imagine it would be adequate to raise an assertion error when it happens. It should not be considered normal.

JordanYates replied:
The challenge with assertions (in my opinion) is that in many applications the assertions aren't compiled in, and are in fact discouraged by the build system outside of tests.
https://github.com/zephyrproject-rtos/zephyr/blob/65f63622e273c19883b866d71a13e8f18b7d5c6c/CMakeLists.txt#L2093-L2097

Specifying that the behavior is "raises an assertion" for me means that most of the time the behavior is "not handled".
In the `init` function I have added the assert case, with an error message stating the duplicated ID that (only compiled in when asserts are disabled).

What behavior do you want to see in this test, given that B and G have the same ID (to validate the error is printed)?

rodrigopex replied:
I like this [text](https://interrupt.memfault.com/blog/defensive-and-offensive-programming) to explain my fondness for assertions. I wouldn't say I like to start developing using defensive programming. The assertion will "guarantee" that developers would not have simple defects in their code. I don't see the message about asserts as "only for tests." It is for development and not production. I rely a lot on assertions on zbus to give developers good and quick feedback about something they are doing incorrectly or inappropriately. A duplicated ID is a case the developer must be aware of as soon as possible to fix, and don't waste time trying to solve some weird behavior because of that. 

> What behavior do you want to see in this test, given that B and G have the same ID (to validate the error is printed)?

I am not sure you can test the assertion results in the init function. For the regular zbus functions, I have created the `ZBUS_ASSERT` that returns `-EFAULT` during tests when the `ZBUS_ASSERT_MOCK` is enabled. In that case, you could return it when a duplication was detected. But inside the init function, I don't know if there is a way.

JordanYates replied:
The assert exists in the `init` function, and if assertions are disabled there is the error log. Both perform the task you desire, quick feedback for development.
I'm not arguing assertions == bad, simply that given the test contains duplicated IDs, we have to choose "something".

* Test `zbus_chan_from_id` with the duplicated ID 
* Don't test `zbus_chan_from_id` with the duplicated ID

If we are testing it with the duplicated ID, we need to choose what it should be returning.

* `NULL` (requires iterating over the entire channel list in every call to detect duplicates)
* First matching channel (no overhead in the "normal" case, minimally wrong if duplicate ID's do exist)

I chose to test it and return the first matching as it is lower overhead for the happy path.
If you think a different choice should be made for the test, I am happy to hear your reasoning.

This is an orthogonal discussion to the assert however, as that is already implemented.

rodrigopex replied:
I see. We need to test that. You are doing that ok. The only attention point is that you are not covering the assert code portion with tests. Are you planning to change some samples to add the ID feature? If so, you could have a version of that with duplicated IDs and try to capture the ASSERTION error string with the `harness: console`. What do you think?

rodrigopex replied:
@JordanYates?

JordanYates replied:
What is the "assert code portion"? This PR does not introduce any new usage of `__ASSERT`.

JordanYates commented:
Rebased on `v3.7.0`, extracted most of the common code into a helper macro.
`init_value` and `observers` parameters must be handled in the first macro call, otherwise the parameter expansion goes crazy.

JordanYates commented:
ping @rodrigopex, this will also need a second reviewer to get merged

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -263,6 +267,30 @@ struct zbus_channel_observation {
 
+#define _ZBUS_MESSAGE_NAME(_name) _CONCAT(_zbus_message_, _name)
+
+#define _ZBUS_CHAN_DEFINE(_name, _id, _type, _validator, _user_data)                               \
```
Can you put the `clang-format off and on` to avoid indentation issues? The GitHub action identifies that.

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -345,37 +377,37 @@ struct zbus_channel_observation {
+ * first the highest priority.
+ * @param _init_val The message initialization.
+ */
+#define ZBUS_CHAN_ID_DEFINE(_name, _id, _type, _validator, _user_data, _observers, _init_val)      \
```
To keep consistency, please use `ZBUS_CHAN_DEFINE_WITH_ID`.

rodrigopex reviewed (changes requested):
Only a few changes are needed. Sorry for the delay. So many things are happening from my side. Thanks for contributing.

pdgendt reviewed (approved):
LGTM

Side note, try to prevent rebasing together with other changes, as this makes the GH "Compare" button useless.

JordanYates commented:
> LGTM
> 
> Side note, try to prevent rebasing together with other changes, as this makes the GH "Compare" button useless.

Interesting, I have always tried to do things in one rebase to prevent unnecessary CI runs

pdgendt commented:
> Interesting, I have always tried to do things in one rebase to prevent unnecessary CI runs

Was the rebase necessary for some CI issues? Otherwise the rebase wasn't even needed, I guess.

rodrigopex commented:
@JordanYates, could you please add documentation for that feature in a separate PR? I noticed it was not included in this PR.
